src/eth/TransactionManager.js

```diff
@@ -149,12 +154,47 @@ export default class TransactionManager extends PublicService {
     return minePromise;
   }
 
-  async _buildTransactionOptions(data) {
+  async _buildTransactionOptions(options, contract, method, args) {
+    const nonce = await this.get('nonce').getNonce();
+
+    if (contract && !options.gasLimit) {
+      await this._getGasLimit(options, contract, method, args, nonce);
+    }
+
     return {
+      ...options,
       ...this.get('web3').transactionSettings(),
-      ...data,
-      nonce: await this.get('nonce').getNonce()
+      nonce: nonce
+    };
+  }
+
+  async _getGasLimit(options, contract, method, args, nonce) {
+    let transaction = {};
+    let data = contract.interface.functions[method](...args).data;
+    let proxyAddress;
+
+    if (options.dsProxy) {
+      proxyAddress = await this.get('proxy').currentProxy();
+      const proxy = this.get('proxy').getUnwrappedProxyContract(proxyAddress);
+      data = proxy.interface.functions['execute'](contract.address, data).data;
+    }
+
+    if (options.value) {
+      transaction.value = options.value;
+    }
+
+    transaction = {
+      from: this.get('web3').currentAddress(),
+      nonce: nonce,
+      to: options.dsProxy ? proxyAddress : contract.address,
+      data: data,
```

Comment:
reminder that you can use the shorthand syntax whenever the key and variable name are the same